src/linker/Linker/EarlyMarker.cs

```diff
@@ -0,0 +1,43 @@
+using Mono.Cecil;
+using System.Diagnostics;
+
+namespace Mono.Linker
+{
+	class EarlyMarker : IMarker
```

Comment:
This looks like unnecessary abstraction, why not to simply have 4 methods inside ResolveFromXmlStep

Reply:
This is so that ResolveFromXml logic can run either early (to mark things before MarkStep calls Initialize*) or during MarkStep (calling the full Mark* logic). I might be able to get rid of the Initialize* logic completely - I am looking into it.

Reply:
Thinking more about this, I think it makes sense to keep this abstraction. If we were to call Mark* logic directly from ResolveFromXmlStep, it would "violate" the pipeline's order of execution, resulting in MarkStep logic called before MarkStep is executing. We could fix that by running the ResolveFromXmlSteps during MarkStep - but then we just end up with a new mini-pipeline of XML steps that get processed during MarkStep.

Another alternative is to never call MarkStep logic from ResolveFromXmlStep (instead only marking things using Annotations) - but this would mean separately keeping track of partially marked things that need to be fully marked later during MarkStep. We would need another HashSet or similar in Annotations to track everything marked from XML but not yet fully marked, which seems a bit wasteful.

Reply:
Add some pipeline concept?

Reply:
This has changed to use the mechanism in https://github.com/mono/linker/pull/1768.